### ISSUE TITLE
server/db: store match secret

### ISSUE THREAD

**Issue by buck54321** [decred/dcrdex#216]
Related to #209, see [previous conversation](https://github.com/decred/dcrdex/pull/209#discussion_r392240227).

The secret is part of the `redeem` and `redemption` serializations, so should be stored.

**Comment by chappjc** [comment]
OK.  For the record, the reason the secret is now relayed by the server is to facilitate SPV clients whose wallets will not have the counterparty's swap contract. Is that right?  I'm still not sure how the client's are supposed to audit the contract if they cannot pull the transaction from the network independent of the server.  Anyway, @jrick said that the wallet will have the transaction if the redeem script is imported, so perhaps that is another way to approach this.

**Comment by buck54321** [comment]
> the reason the secret is now relayed by the server is to facilitate SPV clients whose wallets will not have the counterparty's swap contract. Is that right?

Not just SPV. All clients use the secret. You may be able to import the script, but I don't want to rely on methods like "importscript" whose availability will vary widely, especially since there are no wallet-related scripts in the redemption. Will all assets be able to force their wallets to index non-wallet related transactions?

> I'm still not sure how the clients are supposed to audit the contract if they cannot pull the transaction from the network independent of the server.

Any SPV implementation would need to figure out how to implement a `FindRedemption` method.

**Comment by chappjc** [comment]
I got side tracked on contract auditing, sorry, but I'm a little fuzzy on how the client negotiation currently handles both auditing of the contract and secret extraction from the redemption, despite having just reviewed all that.

1. Contract auditing. The participant/taker needs to check the contract script for the secret hash, lock time, and redeeming addresses, and check the contract output amount.  The client has to be able to get this info without the server's help, or at least with just the contract txid from the server to spare the client from having to search for it, and this means being able to pull a transaction that the wallet may not index.
2. Secret extraction from the redemption. The participant must be able to pull (and possibly find as you point out) the initiator's redeem txn.  But if the server at least provides the redeem txid and the client can pull this redeem txn, then they can extract the secret from the input's signature script.  Unlike with the contract, where the client really must be able to get the raw transaction for themselves, this can be skipped by just checking the hash of the secret against the known secret hash from (1.).

This is why I'm fixating on the contract auditing, because that is required to not trust the server to provide the correct secret hash.  Sorry if I'm being particularly dense about something we've already discussed, but I have a hard time understanding how the contract auditing is achieved trustlessly if the client can't pull arbitrary txns (or import the redeem script).
